spikeinterface/curation/auto_merge.py

```diff
@@ -0,0 +1,264 @@
+    num_spikes = np.array(list(sorting.get_total_num_spikes().values()))
+    to_remove = num_spikes < minimum_spikes
+    pair_mask[to_remove, :] = False
+    pair_mask[:, to_remove] = False
+
+    # STEP 2 : remove contaminated auto corr
+
+    # STEP 3 : unit positions are estimated rougtly with channel
+    chan_loc = we.recording.get_channel_locations()
+    unit_max_chan = get_template_extremum_channel(we, peak_sign=peak_sign, mode="extremum", outputs="index")
+    unit_max_chan = list(unit_max_chan.values())
+    unit_locations = chan_loc[unit_max_chan, :]
+    unit_distances = scipy.spatial.distance.cdist(unit_locations, unit_locations, metric='euclidean')
+    pair_mask = pair_mask & (unit_distances <=maximum_distance_um)
+    
+    print('Will check ', np.sum(pair_mask), 'pairs on ', pair_mask.size)
+    
+    # Step 4 : potential auto merge by correlogram
+    correlograms, bins = compute_correlograms(sorting, window_ms=window_ms, bin_ms=bin_ms, method='numba')
+    corr_diff = compute_correlogram_diff(sorting, correlograms, bins,
+                                    correlogram_low_pass=correlogram_low_pass,  adaptative_window_threshold=adaptative_window_threshold, 
+                                    pair_mask=pair_mask)
+    ind1, ind2 = np.nonzero(corr_diff  < corr_thresh)
+    potential_merges = list(zip(unit_ids[ind1], unit_ids[ind2]))
+    print(potential_merges)
+    
+    # step 5 : check if potential merge with CC also have template similarity
+    # TODO
+    
+    # step 6 : validate the potential merges with CC increase the contamination quality metrics
+    # TODO
+    
+    return potential_merges, corr_diff
+
+
+def compute_correlogram_diff(sorting, correlograms, bins,  correlogram_low_pass=800., adaptative_window_threshold=0.5,
+            pair_mask=None):
+    """
+    Original author: Aurelien Wyngaard ( lussac)
+    
+    Parameters
+    ----------
+    sorting: BaseSorting
+        The sorting object
+    correlograms: array 3d
+        The 3d array containing all cross and auto corrlogram
+    bins: array
+        Bins of the correlograms
+    correlogram_low_pass: float
+        low pass filter on correlogram for smoothing
+    adaptative_window_threshold
+        
+    pair_mask: None or boolean array
+        A bool matrix of size (num_units, num_units) to select
+        which pair to compute.
+    
+    Returns
+    -------
+    corr_diff
+    """
+    bin_ms = bins[1] - bins[0] 
+    
+    unit_ids = sorting.unit_ids
+    n = len(unit_ids)
+    
+    if pair_mask is None:
+        pair_mask = np.ones((n, n), dtype='bool')
+    
+    
+    # Index of the middle of the correlograms.
+    m = correlograms.shape[2] // 2
+    
+    num_spikes = sorting.get_total_num_spikes()
+    
+    # smooth correlogram by low pass filter
+    b, a = scipy.signal.butter(N=2, Wn = correlogram_low_pass / (1e3 / bin_ms /2), btype='low')
+    correlograms_smooth = scipy.signal.filtfilt(b, a, correlograms, axis=2)
```

Comment:
I believe you didn't remove any peaks at t=0 for the cross-correlogram before smoothing. Is that intentional?

This means that if 2 units are very similar but share some spikes, there will be a peak in the auto-correlogram (which will be smoothed) and that will show up as a difference later on.

Reply:
Can you send the line in your original code doing this ?

Reply:
https://github.com/BarbourLab/lussac/blob/a386ae6073aac7b18187e0cf81d4b8cad9de166b/postprocessing/merge_units.py#L164
There's probably a way to vectorize it to remove the double for loop.